src/storage/garbage_collector.cpp

```diff
@@ -129,54 +243,279 @@ void GarbageCollector::ProcessDeferredActions() {
+        ReadUndoRecord(start_record, next, &interval_length);
+      }
+      // Update curr
+      curr = curr->Next();
+    } else {
+      if (interval_length > 1) {
+        // Create the undo record by a second traversal through the records to be compacted.
+        UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+        // Copy the varlen attributes of the compacted record
+        CopyVarlen(compacted_undo_record);
+        // Link the compacted record to the version chain.
+        LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+      }
+      // Set interval length to 0
+      EndCompaction(&interval_length);
+      // Begin compaction for the next interval
+      BeginCompaction(&start_record, curr, next, &interval_length);
+      // curr was not claimed in the previous iteration, so possibly someone might use it
+      curr = curr->Next();
+    }
     next = curr->Next();
-    // This is a legitimate case where we truncated the version chain but had to restart because the previous head
-    // was aborted.
-    if (next == nullptr) return;
-    if (transaction::TransactionUtil::NewerThan(oldest, next->Timestamp().load())) break;
-    curr = next;
-  }
-  // The rest of the version chain must also be invisible to any running transactions since our version
-  // is newest-to-oldest sorted.
+  }
+
+  if (interval_length > 1) {
+    // Create the undo record by a second traversal through the records to be compacted.
+    UndoRecord *compacted_undo_record = CreateUndoRecord(start_record, next);
+    // Copy the varlen attributes of the compacted record
+    CopyVarlen(compacted_undo_record);
+    // Link the compacted record to the version chain.
+    LinkCompactedUndoRecord(start_record, &curr, next, compacted_undo_record);
+  } else if (interval_length == 1) {
+    if (transaction::TransactionUtil::NewerThan(active_txns->back(), curr->Timestamp().load())) {
+      // Exited out of previous loop such that the last record is not visible to any transaction
+      UnlinkUndoRecordVersion(start_record->Next());
+      start_record->Next().store(nullptr);
+    }
+  }
+  // Set interval length to 0
+  EndCompaction(&interval_length);
+
+  // active_trans_iter ends but there are still elements in the version chain. Can GC everything below
   curr->Next().store(nullptr);
+  while (next != nullptr) {
+    // Unlink next
+    UnlinkUndoRecordVersion(next);
+    // Move next pointer ahead
+    next = next->Next();
+  }
+}
 
-  // If the head of the version chain was not committed, it could have been aborted and requires a retry.
-  if (curr == version_ptr && !transaction::TransactionUtil::Committed(version_ptr->Timestamp().load()) &&
-      table->AtomicallyReadVersionPtr(slot, accessor) != version_ptr)
-    TruncateVersionChain(table, slot, oldest);
+void GarbageCollector::UnlinkUndoRecordVersion(UndoRecord *const undo_record) {
```

Comment:
It might be a good idea to check whether `undo_record` is a `nullptr` here, because It seems that not all callers of this method do that check

Reply:
The code logic which calls this function guarantees it cannot be NULL. However, it might be nice to add an assertion.